### DISCUSSION TITLE
Define a required body parameter that can explicitly be set to "null"

### DISCUSSION THREAD

**Discussion by hrist0stoichev** [fastapi/fastapi#12789] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Body
from pydantic import BaseModel

app = FastAPI()

class Model(BaseModel):
    param1: Annotated[str | None, Body(...)]

@app.post("/test")
def test(model: Model):
    return model

@app.post("/test2")
def test2(param1: Annotated[str | None, Body(..., embed=True)]):
    return param1
```

### Description

I want to make the `param1` parameter required but also to be able to pass `null`. For the first endpoint, `test`, it works but it doesn't for `test2` where explicitly passing `null` results in `value_error.missing` error.

```
$ curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test
{"detail":[{"loc":["body"],"msg":"field required","type":"value_error.missing"}

$ curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test --data '{"param1": null}'
{"param1":null}

vs

$ curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test2
{"detail":[{"loc":["body"],"msg":"field required","type":"value_error.missing"}

$ curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test2 --data '{"param1": null}'
{"detail":[{"loc":["body","param1"],"msg":"field required","type":"value_error.missing"}]}
```

### Operating System

macOS

### FastAPI Version

0.99.1

### Pydantic Version

1.10.18

### Python Version

3.10.11

**Comment by YuriiMotov** [comment]
The behavior is really inconsistent.

A couple of remarks:
1) you don't need to use `...` with `Annotated`. To make parameter required just don't specify default value
```
class Model(BaseModel):
    param1: Annotated[str | None, Body()]
```
2) The way you defined your second endpoint it will expect you pass `param1` as just string, not dict with the field `param1`.
To make second endpoint accept parameter the same way it's defined in the first endpoint, you should use `Body(embed=True)`.

But, even for this code it works differently, even though schemas are the same:

```
from typing import Annotated

from fastapi import Body, FastAPI
from pydantic import BaseModel

app = FastAPI()

class Model(BaseModel):
    param1: Annotated[str | None, Body()]

@app.post("/test")
def test(model: Model):
    return model

@app.post("/test2")
def test2(param1: Annotated[str | None, Body(embed=True)]):
    return param1

```

**Reply by hrist0stoichev** [reply]
Regarding 1 - as I've mentioned, I want the parameter to be required (so one must pass it) but also allow for `null` to be passed. For this, `...` is required. Your example doesn't allow `null` to be passed.

Regarding 2 - indeed, I have forgotten to add `embed=True` but for some reason this didn't make any difference and I could still pass param1 in a dict. Still, I tried changing the definition to `param1: Annotated[str | None, Body(..., embed=True)]` but this still didn't allow me to pass explicit `null`.

```
$ curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test2 --data '{"param1": null}'
{"detail":[{"loc":["body","param1"],"msg":"field required","type":"value_error.missing"}]}
```

I will edit my original question to add `embed=True` so that other people don't get confused. Thanks.

**Reply by YuriiMotov** [reply]
Regarding 1: You can just check it, it works as you need without `...`:

```
from typing import Annotated

from fastapi import Body, FastAPI
from pydantic import BaseModel

app = FastAPI()

class Model(BaseModel):
    param1: Annotated[str | None, Body()]

@app.post("/test")
def test(model: Model):
    return model
```

```
curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test
# {"detail":[{"type":"missing","loc":["body"],"msg":"Field required","input":null}]}

curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test --data '{"param1": null}'
# {"param1":null}
```

From the docs:

> Remember that in most of the cases, when something is required, you can simply omit the default, so you normally don't have to use `...`.

`...` is mandatory when you are using non-annotated style to achieve such behavior (can accept `null` but required):
```
class Model(BaseModel):
    param1: str | None = Body(...)
```

**Reply by hrist0stoichev** [reply]
OK, I don't know how I've tested... I tested again and this works without the `...` as you suggested.

Still, my main problem is with point 2.

**Reply by hrist0stoichev** [reply]
Actually, your curl command is wrong as you are not passing a body at all, try
```
curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test --data '{}'
```
And you will see that `...` is indeed required.

**Reply by YuriiMotov** [reply]
Actually, it was the command from your initial question.

> $ curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test
> {"detail":[{"loc":["body"],"msg":"field required","type":"value_error.missing"}

And, I've already tried 
```curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test --data '{}'```
and it returns 
```{"detail":[{"type":"missing","loc":["body","param1"],"msg":"Field required","input":{}}]}```
as expected since you don't pass `null` explicitly.

If you execute
```curl -X POST -H "Content-Type: application/json" http://0.0.0.0:8000/test --data '{"param1": null}'```
you will see that it  accepts `null` value.

**Reply by YuriiMotov** [reply]
And, there is no problem with point 2.
`Body(embed=True)` works. Notice that responses are different for `Body()` and `Body(embed=True)`:

```
{"detail":[{"type":"string_type","loc":["body"],"msg":"Input should be a valid string","input":{"param1":null}}]}
```

VS

```
{"detail":[{"type":"missing","loc":["body","param1"],"msg":"Field required","input":null}]}
```

But I agree that there is highly likely a bug that causes inconsistent behavior when you use `param1: Annotated[str | None, Body(embed=True)]` instead of creating a model

**Comment by redb0** [comment]
I also ran into the problem of not being able to send `null` when using `Body`.

**Comment by YuriiMotov** [comment]
Created a self-descriptive code example with test:

```python
from typing import Annotated
from fastapi import FastAPI, Body
from fastapi.testclient import TestClient
from pydantic import BaseModel
import pytest

app = FastAPI()


class Model(BaseModel):
    param1: Annotated[str | None, Body()]


# The following endpoints have identical request body schemas, but behave differently

@app.post("/model")
def tst(model: Model):
    return model

@app.post("/body-embed")
def tst2(param1: Annotated[str | None, Body(embed=True)]):
    return param1


# Test

@pytest.mark.parametrize("endpoint", ["/model", "/body-embed"])
def test_send_null(endpoint: str):
    """
    Send request body that contains null value for param1.
    Test passes for "/model" endpoint but fails with error code 422 for "/body-embed" endpoint.
    """

    client = TestClient(app)
    response = client.post(endpoint, json={"param1": None})
    assert response.status_code == 200, response.text
    assert response.json() == {"param1": None}


# test_send_null[/body-embed] - AssertionError:
# {"detail":[{"type":"missing","loc":["body","param1"],"msg":"Field required","input":null}]}
```

**Comment by andreasoprani** [comment]
Just encountered this myself, seems like a bug to me, any chance of this being fixed?
